Automattic/woocommerce-payments
# Update the get_repair_data_for_wcpay_items() function to no longer use tax rates and price IDs

Fixes #3173 

#### Changes proposed in this Pull Request

The `get_repair_data_for_wcpay_items()` function is responsible for making sure that the upcoming invoice in Stripe matches the current state of the subscription in WC. 

Prior to this PR it updated products using price IDs and used tax rates. Because of the changes in #3177 and #3176, we no longer use price IDs or tax rates and therefore the `get_repair_data_for_wcpay_items()` needs to make sure the upcoming invoice item's price data matches the current subscription's price data.

The changes in this PR makes sure: 

1. Each WCPay invoice line item (products, fee, shipping, and taxes) all match the line items in WC. 
2. Line item products that have a mismatch in quantity will update the WCPay Subscription to match the quantity in WC.
3. Any item (product, tax, shipping or fee), with a mismatch in line item subtotal, will also lead to an update in WCPay to match the WC subscription. 

#### Testing instructions

#### Set up 

* Make sure you have WCPay Dev and billing clocks enabled.
* Make sure you disable WC Subscriptions.
* Enable shipping and have at least 1 shipping method.
* Enable taxes and have at least 1 applicable tax rate.
* Make sure you're listening to WC Payments webhooks

#### To test

1. Place a subscription in your cart.
3. Purchase the subscription product using WCPay.
2. Go to the edit subscription screen and enable test mode (billing clocks).
3. View the billing clock-enabled subscription in the Stripe dashboard.
4. Go into the database and tweak some of the subscription's quantities and item totals or use add the code snippet below. 
5. Progress the billing clock to both `invoice.upcoming` and `invoice.created`. 
6. View the subscription in Stripe. You should see a subscription now reflects the subscription in WC. 
    - Quantities match
    - Subtotals match
    - The order totals matches
7. Process the invoice as paid via the billing clock UI.
8. Compare the **Payments > Transactions** invoice amount matches the WC order.

-------------------

#### Code snippet
Add this snippet of code to your site to randomise the item totals for you automatically.

```php
add_action( 'init', function() {
	$subscription = wcs_get_subscription( 0 ); // Replace this with your subscription ID.

	if ( ! $subscription || $subscription->meta_exists( '_randomised_item_totals' ) ) {
		return;
	}

	$subscription->update_meta_data( '_randomised_item_totals', 'true' );

	foreach ( $subscription->get_items() as $item ) {
		$coefficient = wp_rand( 80, 300 ) / 100;

		$item->set_quantity( wp_rand( 2, 20 ) );
		$item->set_subtotal( $item->get_subtotal() * $coefficient );
		$item->set_total( $item->get_total() * $coefficient );

		$item->save();
	}

	$subscription->calculate_totals( true );
	$subscription->save();
} );
```

#### Screenshots

<img width="794" alt="Screen Shot 2021-10-21 at 6 01 10 pm" src="https://user-images.githubusercontent.com/8490476/138236257-fab32c78-555a-4091-80a1-3086336028f7.png">

Original order for 4 x ~ $10 product

<img width="1058" alt="Screen Shot 2021-10-21 at 5 58 26 pm" src="https://user-images.githubusercontent.com/8490476/138236326-60feceb5-9362-4aea-8ccc-6b7ff0aa6ed3.png">

After I set up the billing clock I randomised the totals. It is now ~$8000

<img width="1268" alt="Screen Shot 2021-10-21 at 5 58 45 pm" src="https://user-images.githubusercontent.com/8490476/138236346-88c8217e-a27a-441d-837a-c6eb6188483e.png">

<img width="1065" alt="Screen Shot 2021-10-21 at 5 59 58 pm" src="https://user-images.githubusercontent.com/8490476/138236368-9cb6fb5e-0398-492e-9c02-cb81f33ac6fc.png">


**Post merge**
